File: Neetcode/MergeKSortedLL.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
from queue import PriorityQueue
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        if not lists:
            return None
        pq = PriorityQueue()
        knodes = len(lists)
        if knodes < 2:
            return lists[0]

        nodedict = dict()

        for k in range(knodes):
            if lists[k]:
                nodedict[k] = lists[k]
                pq.put((lists[k].val, k))

        dummy = ListNode(0)
        returnnode = dummy

        while not pq.empty():
            smallest, key = pq.get()
            node = nodedict[key]
            dummy.next = node
            dummy = dummy.next
            if nodedict[key].next is not None:
                nodedict[key] = nodedict[key].next
                pq.put((nodedict[key].val, key))

        return returnnode.next

File: Neetcode/MergeKSortedLL.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        struct custom{
            bool operator()(ListNode* a, ListNode* b){
                return a->val > b->val;
            }
        };
        priority_queue<ListNode*, vector<ListNode*>, custom> pq;
        if (lists.empty()){
            return nullptr;
        }
        if (lists.size() == 1){
            return lists[0];
        }
        for (auto l : lists){
            if (l != nullptr)
                pq.push(l);
        }
        if (pq.empty()){
            return nullptr;
        }
        ListNode* head = pq.top();
        ListNode* top = head;
        while (!pq.empty()){
            top = pq.top();
            pq.pop();
            if (top->next != nullptr){
                pq.push(top->next);
            }
            if (!pq.empty()){
                top->next = pq.top();
            }
        }
        return head;
    }
};
